File: neo4j_util_go/neo4j_interface.go
package neo4j_util_go

import (
	"fmt"
	"log"
	"os"
	"time"

	"github.com/joho/godotenv"
	"github.com/neo4j/neo4j-go-driver/v5/neo4j"
)

// Neo4jInterface provides an interface to interact with Neo4j database
type Neo4jInterface struct {
	driver neo4j.Driver
}

// NewNeo4jInterface initializes a new Neo4j connection
func NewNeo4jInterface() (*Neo4jInterface, error) {
	// Load environment variables
	err := godotenv.Load()
	if err != nil {
		log.Println("Warning: No .env file found. Using system environment variables.")
	}

	// Get Neo4j connection details
	uri := os.Getenv("NEO4J_URI")
	if uri == "" {
		uri = "neo4j://localhost:7687" // Default URI
	}

	username := os.Getenv("NEO4J_USERNAME")
	if username == "" {
		username = "neo4j" // Default username
	}

	password := os.Getenv("NEO4J_PASSWORD")
	if password == "" {
		log.Println("Warning: NEO4J_PASSWORD not set. Using empty password.")
	}

	// Create Neo4j driver
	driver, err := neo4j.NewDriver(
		uri,
		neo4j.BasicAuth(username, password, ""),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create Neo4j driver: %w", err)
	}

	// Verify connection
	err = driver.VerifyConnectivity()
	if err != nil {
		// Close the driver if verification fails
		driver.Close()
		return nil, fmt.Errorf("failed to connect to Neo4j: %w", err)
	}

	fmt.Println("✅ Connected to Neo4j")

	return &Neo4jInterface{
		driver: driver,
	}, nil
}

// Close closes the Neo4j connection
func (n *Neo4jInterface) Close() {
	if n.driver != nil {
		n.driver.Close()
		fmt.Println("Neo4j connection closed.")
	}
}

// ExecuteQuery executes a Cypher query with parameters and returns the result
func (n *Neo4jInterface) ExecuteQuery(query string, params map[string]interface{}) ([]map[string]interface{}, error) {
	session := n.driver.NewSession(neo4j.SessionConfig{
		AccessMode: neo4j.AccessModeWrite,
	})
	defer session.Close()

	// Execute the query
	result, err := session.Run(query, params)
	if err != nil {
		return nil, fmt.Errorf("failed to run query: %w", err)
	}

	// Collect all records
	records, err := result.Collect()
	if err != nil {
		return nil, fmt.Errorf("failed to collect results: %w", err)
	}

	// Convert records to maps
	var resultMaps []map[string]interface{}
	for _, record := range records {
		recordMap := make(map[string]interface{})
		for i, key := range record.Keys {
			recordMap[key] = record.Values[i]
		}
		resultMaps = append(resultMaps, recordMap)
	}

	return resultMaps, nil
}

// ExecuteQueryWithRetry executes a Cypher query with retry mechanism
func (n *Neo4jInterface) ExecuteQueryWithRetry(query string, params map[string]interface{}, maxAttempts int) ([]map[string]interface{}, error) {
	if params == nil {
		params = make(map[string]interface{})
	}

	var lastErr error

	for attempt := 1; attempt <= maxAttempts; attempt++ {
		fmt.Printf("Attempt %d of %d to execute query...\n", attempt, maxAttempts)

		result, err := n.ExecuteQuery(query, params)
		if err == nil {
			fmt.Println("Query executed successfully.")
			return result, nil
		}

		lastErr = err
		fmt.Printf("Error on attempt %d: %v\n", attempt, err)

		if attempt < maxAttempts {
			// Exponential backoff
			waitTime := time.Duration(2*attempt) * time.Second
			fmt.Printf("Waiting %v seconds before retrying...\n", waitTime.Seconds())
			time.Sleep(waitTime)
		}
	}

	return nil, fmt.Errorf("maximum retry attempts reached. Last error: %w", lastErr)
}

File: neo4j_util/neo4j_interface.py
import os
import time
import logging
from neo4j import GraphDatabase
from dotenv import load_dotenv

class Neo4jInterface:
    """Interface to interact with Neo4j database"""
    
    def __init__(self):
        """Initialize the Neo4j connection"""
        # Load environment variables
        load_dotenv()
        
        # Get Neo4j connection details
        uri = os.getenv("NEO4J_URI", "neo4j://localhost:7687")
        username = os.getenv("NEO4J_USERNAME", "neo4j")
        password = os.getenv("NEO4J_PASSWORD", "")
        
        if not password:
            logging.warning("NEO4J_PASSWORD not set. Using empty password.")
        
        try:
            # Create Neo4j driver
            self.driver = GraphDatabase.driver(uri, auth=(username, password))
            
            # Verify connection
            self.driver.verify_connectivity()
            print("✅ Connected to Neo4j")
        except Exception as e:
            raise ConnectionError(f"Failed to connect to Neo4j: {e}")
    
    def close(self):
        """Close the Neo4j connection"""
        if hasattr(self, 'driver'):
            self.driver.close()
            print("Neo4j connection closed.")
    
    def execute_query(self, query, params=None):
        """Execute a Cypher query with parameters and return the result
        
        Args:
            query (str): The Cypher query to execute
            params (dict): Parameters for the query
            
        Returns:
            list: List of record dictionaries
        """
        if params is None:
            params = {}
        
        with self.driver.session() as session:
            result = session.run(query, params)
            records = result.data()  # Convert to a list of dictionaries
            return records
    
    def execute_query_with_retry(self, query, params=None, max_attempts=3):
        """Execute a Cypher query with retry mechanism
        
        Args:
            query (str): The Cypher query to execute
            params (dict): Parameters for the query
            max_attempts (int): Maximum number of retry attempts
            
        Returns:
            list: List of record dictionaries
        """
        if params is None:
            params = {}
        
        last_err = None
        
        for attempt in range(1, max_attempts + 1):
            print(f"Attempt {attempt} of {max_attempts} to execute query...")
            
            try:
                result = self.execute_query(query, params)
                print("Query executed successfully.")
                return result
            except Exception as e:
                last_err = e
                print(f"Error on attempt {attempt}: {e}")
                
                if attempt < max_attempts:
                    # Exponential backoff
                    wait_time = 2 * attempt
                    print(f"Waiting {wait_time} seconds before retrying...")
                    time.sleep(wait_time)
        
        raise Exception(f"Maximum retry attempts reached. Last error: {last_err}") 